ddtrace/internal/writer.py

```diff
@@ -347,9 +347,9 @@ def write(self, spans):
         # Start the AgentWriter on first write.
         # Starting it earlier might be an issue with gevent, see:
         # https://github.com/DataDog/dd-trace-py/issues/1192
-        if self.started is False:
+        if not self.is_alive():
             with self._started_lock:
-                if self.started is False:
+                if not self.is_alive():
```

Comment:
Can you explain why we have to check again if the AgentWriter is alive?

Reply:
Haha this change was accidental!

Reply:
I was just curious why in general we have to check `if self.started is False` again when we've already checked that in the line above 😕

Reply:
Oh sure!

First off, the race condition here is that a thread can not be started more than once (else an exception is thrown). 

The first check is done outside of the lock as an optimization. If `self.started` is `True` then the thread has already started so there's no need to acquire the lock. Now if `self.started` is `False` we don't know if another thread might also have just read the same condition (since there's no lock). So the lock is acquired and then we check again to achieve the mutual exclusion.

Reply:
Ahhh, that makes a lot of sense now. Thanks for the explanation!